### PR TITLE
[MOBILE] implement android widget - v1

### PR DESCRIPTION
**Pull request by Mnoble-19**
#### Summary of Changes (What does this PR do?)

- Create homescreen widget for android application

#### Status of maturity (all need to be checked before merging):

- [x] I've tested this locally
- [x] I consider this code done
- [x] This change ready to hit production in its current state
- [x] The title of the PR states what changed and the related issues number (used for the release note).
- [] I've included issue number in the "Closes #ISSUE-NUMBER" part of the "[What are the relevant tickets?](#what-are-the-relevant-tickets)" section to [link the issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).
- [ ] I've updated corresponding documentation for the changes in this PR.
- [ ] I have written unit and/or e2e tests for my change(s).

#### How should this be manually tested?

- Get this code branch
- Open the `android` directory in android studio.
- Build the application
- check your phone's widget options and select the AirQo widget

#### What are the relevant tickets?

- Closes #<enter issue number here>
- [Jira_card_number]() (If exists)

### CODE REVIEW

File: mobile/android/app/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -1,46 +1,74 @@
+<?xml version="1.0" encoding="utf-8"?>
 <manifest xmlns:android="http://schemas.android.com/apk/res/android"
     package="com.airqo.app">
 
     <uses-permission android:name="android.permission.INTERNET" />
```
Comment (Baalmart) [inline mobile/android/app/src/main/AndroidManifest.xml:5]: @Mnoble-19 , somehow unrelated....in these manifests, isn't this where we are supposed to deal with the permissions issues we currently have where the app stays running in the background after a user has exited it?

**Review by Baalmart — COMMENTED** [review]
Thanks @Mnoble-19 for these works.

Awaiting to see the progress around IOS works, as I see Android is the focal point here. Unless I am mistaken.

Thanks again!

**Comment by Mnoble-19** [comment]
> Thanks @Mnoble-19 for these works.
> 
> Awaiting to see the progress around IOS works, as I see Android is the focal point here. Unless I am mistaken.
> 
> Thanks again!

Yes @Baalmart , this PR is mostly Android, then setting up the data transfer from the main app to the widget, iOS will come in a different pull request :) Stay tuned!

**Comment by Mnoble-19** [comment]
> @Mnoble-19 I do notice some merge conflicts in the PR....also, you will want to add an appropriate label to the PR: is it `work-in-progress `or `ready-to-review`?

It's now ready for review. Was still doing some cleanup.

File: mobile/lib/screens/dashboard/dashboard_view.dart [hunk 2]
```diff
@@ -464,6 +468,123 @@ class _DashboardViewState extends State<DashboardView>
...
+    // };
+
+    // return Future.wait([
+    //   widgetData.forEach((key, value) {
+    //     HomeWidget.saveWidgetData<String>(key, value);
+    //   })
+    // ] as Iterable<Future>)
+    //     .then((value) => value);
+
+    return Future.wait([
```
Comment (Mnoble-19) [inline mobile/lib/screens/dashboard/dashboard_view.dart:518]: Hi @NoahNsimbe , the code below  goes extremely against the DRY principle. I tried to use a forEach loop and map above but it wasn't passing the data as expected thus i opted for what's below, kindly share thoughts

File: mobile/android/app/src/main/kotlin/com/airqo/app/AirQoCircularWidget.kt [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+                        ColorRange(13, 35, R.drawable.yellow_circle, 0xffA8A800.toInt()),
+                        ColorRange(36, 55, R.drawable.orange_circle, 0xffB86000.toInt()),
+                        ColorRange(56, 150, R.drawable.red_circle, 0xffB80B00.toInt()),
+                        ColorRange(151, 250, R.drawable.purple_circle, 0xff8E00AC.toInt()),
+                        ColorRange(
+                            Int.MAX_VALUE,
+                            Int.MAX_VALUE,
+                            R.drawable.maroon_circle,
+                            0xffDBA5B2.toInt()
+                        )
```
Comment (NoahNsimbe) [inline mobile/android/app/src/main/kotlin/com/airqo/app/AirQoCircularWidget.kt:61]: Its better to transfer this logic to flutter and use existing functionality.

File: mobile/android/app/src/main/kotlin/com/airqo/app/AirQoHomeScreenWidget.kt [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+                        ColorRange(36, 55, R.drawable.orange_circle, 0xffB86000.toInt()),
+                        ColorRange(56, 150, R.drawable.red_circle, 0xffB80B00.toInt()),
+                        ColorRange(151, 250, R.drawable.purple_circle, 0xff8E00AC.toInt()),
+                        ColorRange(
+                            Int.MAX_VALUE,
+                            Int.MAX_VALUE,
+                            R.drawable.maroon_circle,
+                            0xffDBA5B2.toInt()
+                        )
+                    )
```
Comment (NoahNsimbe) [inline mobile/android/app/src/main/kotlin/com/airqo/app/AirQoHomeScreenWidget.kt:72]: DRY

File: mobile/android/app/src/main/res/layout/air_qo_circular_widget.xml [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+            android:paddingBottom="10dp"
+            android:text="@string/pmScale"
+            android:textAppearance="@style/pmScaleCircle" />
+
+        <TextView
+            android:id="@+id/circular_pm_value"
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:layout_centerInParent="true"
+            android:fontFamily="@font/roboto_mono_medium"
```
Comment (NoahNsimbe) [inline mobile/android/app/src/main/res/layout/air_qo_circular_widget.xml:33]: To reduce the app download size, Could we try exploring usage of downloadable fonts. 

https://github.com/android/user-interface-samples/tree/main/DownloadableFonts

File: mobile/android/app/src/main/res/values/strings.xml [hunk 6]
```diff
@@ -1,4 +1,19 @@
...
+    <string name="pmValue">40</string>
+    <string name="pmUnit">µg/m<sup>3</sup></string>
+    <string name="forecastValue1">22</string>
+    <string name="time1">4pm</string>
+    <string name="forecastValue2">23</string>
+    <string name="time2">5pm</string>
+    <string name="forecastValue3">24</string>
+    <string name="time3">6pm</string>
+    <string name="refresh_icon">Refresh icon</string>
+    <string name="location_icon">Location icon</string>
```
Comment (NoahNsimbe) [inline mobile/android/app/src/main/res/values/strings.xml:18]: This looks like hard coded data. Do we plan to release it to production??

File: mobile/android/app/src/main/res/values/styles.xml [hunk 7]
```diff
@@ -16,4 +16,68 @@
     <style name="NormalTheme" parent="@android:style/Theme.Light.NoTitleBar">
         <item name="android:windowBackground">@color/whiteColor</item>
     </style>
+
+    <style name="wandegeya">
```
Comment (NoahNsimbe) [inline mobile/android/app/src/main/res/values/styles.xml:20]: We might need to assign another name to this style. `wandegeya` sounds like a place

File: mobile/lib/screens/dashboard/dashboard_view.dart (hunk 2 above)
Comment (NoahNsimbe) [reply mobile/lib/screens/dashboard/dashboard_view.dart:518]: We shall review during a peer programming session. We shall create a class for widget data

File: mobile/android/app/src/main/res/values/strings.xml [hunk 8]
```diff
@@ -1,4 +1,19 @@
 <?xml version="1.0" encoding="utf-8"?>
 <resources>
-    <string name="app_name">Default App Name</string>
+    <string name="app_name">AirQo</string>
+    <string name="app_widget_description">This is the AirQo application homescreen widget</string>
+    <string name="location">Wandegeya</string>
```
Comment (NoahNsimbe) [inline mobile/android/app/src/main/res/values/strings.xml:5]: We can replace this with an African city.  @Mnoble-19

File: mobile/android/app/src/main/AndroidManifest.xml (hunk 1 above)
Comment (Baalmart) [reply mobile/android/app/src/main/AndroidManifest.xml:5]: Thanks @Mnoble-19 ....so is this issue only in IOS?

**Comment by Baalmart** [comment]
Hi @Mnoble-19 , the PR has merge conflicts, please address them. And get your teammates to review the PR.

**Comment by Baalmart** [comment]
@Mnoble-19 , To minimise confusion, please change the label for this PR to something more appropriate and perhaps change it to a DRAFT PR. Assuming that you put a pause on it. Thanks!

**Review by BenjaminSsempala — APPROVED** [review]
LGTM

**Pull request merged**